internal/dataplane/kongstate/kongstate.go

```diff
@@ -177,18 +179,96 @@ func (ks *KongState) FillConsumersAndCredentials(
 				}
 				credConfig[k] = string(v)
 			}
-			credTags := util.GenerateTagsForObject(secret)
-			if err := c.SetCredential(credType, credConfig, credTags); err != nil {
+			tags := util.GenerateTagsForObject(secret)
+			cred, err := c.SetCredential(credType, credConfig, tags)
+			if err != nil {
 				pushCredentialResourceFailures(
 					fmt.Sprintf("failed to provision credential: %v", err),
 				)
 				continue
 			}
+			credentialCollection.collectCredential(cred, secret, consumer)
 		}
 
 		consumerIndex[consumer.Namespace+"/"+consumer.Name] = c
 	}
 
+	// remove consumers with conflicting credentials in generated Kong state.
+	purgeConsumerWithConflictingCredential := func(
+		kc *kongv1.KongConsumer,
+		msg string,
+	) {
+		if kc == nil {
+			return
+		}
+		consumerKey := kc.Namespace + "/" + kc.Name
+		_, ok := consumerIndex[consumerKey]
+		if !ok {
+			return
+		}
+		failuresCollector.PushResourceFailure(
+			fmt.Sprintf("Consumer has conflicting credentials: %s", msg),
+			kc,
+		)
+		delete(consumerIndex, consumerKey)
+	}
+
+	conflictingKeyAuths := getKeyAuthsConflictingOnKey(credentialCollection.KeyAuths)
```

Comment:
I believe we could make this more performant if we build the indices for credentials when we collect them. We could also make the pieces responsible for detecting conflicts to be encapsulated in a separate struct and a separate file. Please see my suggestion: https://github.com/Kong/kubernetes-ingress-controller/compare/fix/reject_duplicate_keys_in_translator...fix/reject_duplicate_keys_in_translator-suggestion

Reply:
Cherry-picked the commit in your suggestion branch.